Add tests and support S3 with tensorboard-notf

orionr opened the pull request:
As a part of building out tensorboard-notf, we need support for S3 (and our internal filesystems) with the tensorboard-notf build.

This is a continuation of https://github.com/tensorflow/tensorboard/pull/1418 that adds robust test coverage and S3 support.

Tested S3 with

```
pip install boto3
pip install moto
vi ~/.aws/credentials # setup per https://stackoverflow.com/questions/33297172/boto3-error-botocore-exceptions-nocredentialserror-unable-to-locate-credential
bazel build tensorboard:tensorboard-notf --verbose_failures
./bazel-bin/tensorboard/tensorboard-notf --logdir ~/local/sample_runs
./bazel-bin/tensorboard/tensorboard-notf --logdir s3://escapes/tensorboard-notf/sample_runs/runs/Jul05_11-15-47_limichael-mbp
./bazel-bin/tensorboard/tensorboard-notf --logdir s3://escapes/tensorboard-notf/sample_runs # Note that this will be slow to load
```

Also can run the following tests

```
bazel test //tensorboard/compat/proto:proto_test --test_output=errors
bazel test //tensorboard/compat/tensorflow_stub:gfile_test --test_output=errors
bazel test //tensorboard/compat/tensorflow_stub:gfile_s3_test --test_output=errors
```

cc @nfelt and @lanpa

orionr commented:
Still have some issues here - stay tuned.

orionr commented:
@nfelt (and others) can we get a review on this? Much appreciated.

orionr commented:
@nfelt, great feedback. I agree that simplifying to a read-only (or near read-only) API makes a lot of sense. That being said, are you open to landing this with changes before we simplify things further? Let me know. In the meantime, I'll see if I can address the comments above - thanks.

orionr commented on tensorboard/compat/tensorflow_stub/pywrap_tensorflow.py:
```diff
@@ -164,64 +164,77 @@ def crc32c(data):
+      start_offset=0,
+      compression_type=None,
+      status=None,
+      check_crc32c=False
```
This was primarily a performance issues - having the check slowed things down considerably. I forget what the numbers were exactly, though.

orionr commented on tensorboard/compat/tensorflow_stub/gfile.py:
```diff
@@ -37,41 +44,343 @@
+        """Determines whether a path exists or not."""
+        try:
+            return os.path.exists(compat.as_bytes(filename))
+        except errors.NotFoundError:
```
You're right - I should have caught this one. Thanks.

nfelt commented:
From an initial audit, I think the minimal subset of the gfile API needed for basic TB functionality would be implementations of:

- exists()
- glob()
- isdir()
- listdir()
- stat()
- walk()
- read_file_to_string()

Scoping it down to just this would help a lot and I think should still get you what you need - it wouldn't support the projector plugin for now which makes use of gfile.GFile to read file contents, but I'd prefer not to generalize GFile now only to later get rid of it entirely.

The other main concern is testing: if we're adding further code to the stub, it needs test coverage - submitting the stub code without tests was a special case since it was largely derived from TF's code, but not something we can continue doing.  (As an aside, we also still need at least some smoke testing that TB works in no-TF mode, since I realized that a recent change migrated our gfile API calls from tf.gfile to the new tf.io.gfile with different symbol names, so the no-TF build is presumably non-functional right now.)

If we can tackle those two within the PR I'm open to merging, though I still think it may be more efficient overall to identify a better medium-term solution and then migrate directly towards that.

orionr commented:
I agree - unit tests are a big gap right now. If you have any pointers on standard practice, please pass them my way. This will take me a bit longer, but worth the investment.

nfelt commented:
Thanks!  I'd say existing unit tests in the codebase are our best example of standard practice, e.g. something like this test for the io_wrapper functionality: https://github.com/tensorflow/tensorboard/blob/a67fc6c27e5c859eb7135e005eda5259b109638b/tensorboard/backend/event_processing/io_wrapper_test.py

I'm not sure what the best way to test the boto3 API calls would be.  From a quick search I found https://github.com/spulec/moto which looked reasonable at first glance.

orionr commented:
I've cleaned up the GFile interface to be primarily read-only and matching the new `tf.io.gfile` location. I was also able to unify read to one location and am now looking at chunking / buffering. Once I have that I'll push and then work on tests. Thanks for the followup!

lanpa commented:
FYI: some thing you might need to use moto. (`export...` and `os.environment(...`
https://github.com/lanpa/tensorboardX/commit/14e35a19ce28fc29935db8de622281d254702548

orionr commented:
Rebased although it looks like some of the refactors might not have played nice... @nfelt should I pull some of this out of the PR so we can land incrementally? In any case, we now have a robust test suite here.

orionr commented:
All tests pass. @nfelt, can we see how to land this sooner rather than later? Seems like things are changing underneath, so hopefully we can get some good tests in place. Thank you!

nfelt commented on tensorboard/compat/proto/BUILD:
```diff
@@ -48,10 +48,6 @@ py_test(
     srcs = ["proto_test.py"],
     srcs_version = "PY2AND3",
-    tags = [
-        "manual",
```
I've actually already done part 2 above since we needed it for our 1.13 release: https://github.com/tensorflow/tensorboard/pull/1823

So those changes can just be dropped from this PR.

orionr commented on .travis.yml:
```diff
@@ -125,6 +130,8 @@ script:
   - bazel build //tensorboard/...
   - bazel test //tensorboard/...
+  # Run manual S3 test
+  - bazel test //tensorboard/compat/tensorflow_stub:gfile_s3_test
```
This test needs to be manual since it requires boto3 and we likely shouldn't require that everyone install it. @nfelt, thoughts?

orionr commented on tensorboard/compat/tensorflow_stub/io/gfile.py:
```diff
@@ -0,0 +1,620 @@
+            raise NotImplementedError(
+                "mode {} not supported by compat GFile".format(mode))
+        self.filename = compat.as_bytes(filename)
+        self.buff_chunk_size = 16 * 1024 * 1024
```
Oops! Thanks.

orionr commented on tensorboard/compat/tensorflow_stub/io/gfile.py:
```diff
@@ -0,0 +1,620 @@
+        self.length = None
+        self.binary_mode = (mode != 'r')
+
+    def _fill_buffer_to(self, size):
```
Although I like the simplicity of this, I think you will have a problem when `__next__` needs to return a line beyond a single buffer chunk. I'll adjust accordingly.

nfelt replied:
Yes, you're right - `__next__()` should build up a temporary `line` buffer of its own with each `read()` call and return that.  I think it still has the same advantages I mentioned though, even if it's slightly less simple with that correction.

orionr commented on tensorboard/plugins/audio/BUILD:
```diff
@@ -45,6 +45,27 @@ py_test(
 )
 
+py_test(
+    name = "audio_plugin_test_notf",
```
I think my concern with this was that the tests used TensorFlow for a variety of things, so I didn't want to explicitly require that we didn't have TF installed to run "notf" tests. I agree, though, that this adds a lot of boilerplate. For example, `audio_plugin_test.py` extends `tf.test.TestCase` and uses TensorFlow to write out some sample data. Thoughts on this?

orionr replied:
I'll breakout these tests from the rest of the PR regardless.

orionr commented on tensorboard/compat/tensorflow_stub/pywrap_tensorflow.py:
```diff
@@ -164,64 +164,90 @@ def crc32c(data):
-            self.done_reading = True
+        self.done_reading = True
+
+    def read_and_check(buf, offset, num_bytes):
```
Removing.

orionr commented on tensorboard/compat/tensorflow_stub/io/gfile.py:
```diff
@@ -0,0 +1,620 @@
+                key = o["Key"][len(path):]
+                if key:  # Skip the base dir, which would add an empty string
+                    keys.append(key)
+        return keys
```
Correct - I think these are always `str`.

orionr commented:
@nfelt should have addressed all your concerns here. Let me know if I missed anything. Also, moved plugin tests themselves (so we can discuss more) to https://github.com/tensorflow/tensorboard/pull/1829/ . Thanks.

orionr commented on tensorboard/compat/tensorflow_stub/io/gfile.py:
```diff
@@ -0,0 +1,620 @@
+            raise NotImplementedError(
+                "{} not supported by compat glob".format(filename))
+        if star_i != len(filename) - 1:
+            # Just return empty so we can use glob from directory watcher
```
I can see your point here. However, we want to use this for our internal distributed filesystem as well and doing a check for S3 in GetLogdirSubdirectories wouldn't be sufficient. Can we put a TODO here instead of making the larger change?

orionr commented on tensorboard/compat/tensorflow_stub/pywrap_tensorflow.py:
```diff
@@ -177,51 +183,54 @@ def read(self):
-                # header_crc_calc = masked_crc32c(header_str)
-                # assert header_crc_calc == crc_header[0], \
-                #     'Header crc\'s dont match'
+                # Read the crc32, which is 4 bytes, and check it against
```
Note that we now do the crc32 checks in all cases. If perf is an issue we consider changing in the future.

orionr commented:
@stephanwlee can you test this on Windows? Much appreciated.

orionr commented:
@stephanwlee confirmed this passed on Windows with all changes. @nfelt do you need anything from me to land this? Thanks.

orionr commented:
All changes should have been made! @nfelt please take a look and thanks.

nfelt commented:
Merged!  Thanks for bearing with us.

wchargin commented on tensorboard/compat/tensorflow_stub/io/gfile_s3_test.py:
```diff
@@ -0,0 +1,315 @@
+
+from tensorboard.compat.tensorflow_stub.io import gfile
+
+os.environ.setdefault("AWS_ACCESS_KEY_ID", "foobar_key")
```
The comment says that these overwrite any local keys, but the code
actually keeps the local keys if they’re there, only setting the
placeholders if the user has no keys. Am I missing something? Shouldn’t
these just be

```python
os.environ["AWS_ACCESS_KEY_ID"] = "foobar_key"
os.environ["AWS_SECRET_ACCESS_KEY"] = "foobar_secret"
```
